src/mpfao/3D/tdympfao3D_ts.c

```diff
@@ -324,19 +309,26 @@ PetscErrorCode TDyMPFAOIJacobian_Vertices_3DMesh(Vec Ul, Mat A, void *ctx) {
       if (cell_id_up<0) cell_id_up = cell_id_dn;
       if (cell_id_dn<0) cell_id_dn = cell_id_up;
 
-      dden_dPup = 0.0;
-      dden_dPdn = 0.0;
+      PetscReal dden_dPup = 0.0;
+      PetscReal dden_dPdn = 0.0;
 
       if (cell_id_up>=0) dden_dPup = 0.5*tdy->drho_dP[cell_id_up];
       if (cell_id_dn>=0) dden_dPdn = 0.5*tdy->drho_dP[cell_id_dn];
 
-      for (icol=0; icol<vertices->num_internal_cells[ivertex]; icol++) {
-        cell_id = vertices->internal_cell_ids[vOffsetCell + icol];
+      PetscInt num_int_cells = vertices->num_internal_cells[ivertex];
+      PetscInt up_cols[num_int_cells], dn_cols[num_int_cells];
+      PetscReal up_Jac[num_int_cells], dn_Jac[num_int_cells];
+      for (PetscInt icol=0; icol < num_int_cells; ++icol) {
+        up_cols[icol] = -1;
+        dn_cols[icol] = -1;
+        PetscInt cell_id = vertices->internal_cell_ids[vOffsetCell + icol];
 
-        T = tdy->Trans[vertex_id][irow][icol];
+        PetscReal T = tdy->Trans[vertex_id][irow][icol];
 
+        PetscReal gz;
         ierr = ComputeGtimesZ(tdy->gravity,cells->centroid[cell_id].X,dim,&gz); CHKERRQ(ierr);
 
+        PetscReal Jac;
         if (cell_id_up>-1 && cell_id == cell_id_up) {
```

Comment:
A random thought. If I understand vectorization, I think this code can also be vectorized as the conditionals are based on cell-ids and they will not change during time-stepping. For a given `icol`, the `J` can be written as:
```
J =  aJ^1_up + bJ^2_up + cJ^3_up + dJ^1_dn + eJ^2_dn + fJ^3_dn + J^4
```
where 
- `J^1_up = dden_dPup * ukvr       * TtimesP[irow]` (does not change for `icol`)
- `J^2_up = den       * dukvr_dPup * TtimesP[irow]` (does not change for `icol`)
- `J^3_up = den       * ukvr       * T * dden_dPup*gz ` (does change for `icol`)
- `J^1_dn = dden_dPdn * ukvr       * TtimesP[irow]` (does not change for `icol`)
- `J^2_dn = den       * dukvr_dPdn * TtimesP[irow]` (does not change for `icol`)
- `J^3_dn = den       * ukvr       * T * dden_dPdn*gz ` (does change for `icol`)
- `J^3.  = den       * ukvr       * T` (does change for `icol`)

The coefficients `a`-`f` for various Jacobian terms will be saved during initialization and will be given by:
```
J_up    = 1 x J^1_up + 1 x J^2_up + 1 x J^3_up + 0 x J^1_dn + 0 x J^2_dn + 0 x J^3_dn + 1 x J^4
J_dn    = 0 x J^1_up + 0 x J^2_up + 0 x J^3_up + 1 x J^1_dn + 1 x J^2_dn + 1 x J^3_dn + 1 x J^4
J_other = 0 x J^1_up + 0 x J^2_up + 0 x J^3_up + 0 x J^1_dn + 0 x J^2_dn + 0 x J^3_dn + 1 x J^4
```

Reply:
I've recorded this link in a vectorization-related issue. For now, I'll get this PR merged.